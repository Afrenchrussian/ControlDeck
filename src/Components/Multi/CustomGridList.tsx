import React from "react";
import { GridList, GridListTile, makeStyles, Tooltip, Typography } from "@material-ui/core";
import ButtonBase from "@material-ui/core/ButtonBase";
import { ErrorOutlineTwoTone, PhoneIphone, TabletMac } from "@material-ui/icons";
import { Device, Profile } from "../../Store/Reducers/mainReducer";

/*
    data: can be any of the types listed as only a name and a type is required
          (name : used for id and shown in tile // TESTED , type: used to pick which icon to show )
    onTileClick: when a tile is clicked this call back is called with the index of the tile clicked // TESTED
    checkForWarnings: if defined should return a string containing a warning for any cases, // TESTED
                        callback will be passed the elements data
    //TODO deletable: NOT IMPLEMENTED -- when true will allow any of the tiles to be deleted
    //TODO onDelete: NOT IMPLEMENTED -- called when a tile is clicked to be deleted
 */

interface Props {
    data: Profile[] | Device[] | {name: string, type: number}[],
    onTileClick?: (index: number) => void
    checkForWarnings?: (element: any) => string,
    deletable?: boolean,
    onDelete?: (index: number) => void
}

const useStyles = makeStyles((theme) => ({
    card: {
        width: "180px",
        height: "110px",
        backgroundColor: theme.palette.primary.main,
        "&:hover": {
            boxShadow: "0px 3px 6px #00000029",
            cursor: "pointer"
        }
    },
    text: {
        margin: "auto",
        textTransform: "capitalize",
        color: "white"
    },
    icon_large: {
        fontSize: "55px",
        color: "white",
        margin: "auto"
    },
    icon_small: {
        fontSize: "35px",
        color: "white",
        margin: "auto"
    }
}));

export default function CustomGridList(props: Props) {
    const classes = useStyles();

    const icon = (type: number) => {
        switch (type) {
            case 0:
                return <PhoneIphone className={classes.icon_large}/>;
            case 1:
                return <PhoneIphone className={classes.icon_small}/>;
            case 2:
                return <TabletMac className={classes.icon_large}/>;
            case 3:
                return <TabletMac className={classes.icon_small}/>;
        }
    };

    return (
        <GridList cols={3} style={{ width: "602px", margin: "auto", marginTop: "30px" }}>
            {
                (props.data as {name: string, type: number}[]).map((cur, index) => {
                const warning = (props.checkForWarnings !== undefined && props.checkForWarnings(cur) !== "") ? (
                    <Tooltip id={cur.name + "_warning"} title={props.checkForWarnings(cur)}>
                        <ErrorOutlineTwoTone
                            style={{ color: "yellow", position: "absolute", opacity: 0.6, right: "5px", top: "5px" }}/>
                    </Tooltip>
                ) : <React.Fragment/>;

                return (
                    <GridListTile cols={1} key={index} style={{ height: "135px" }}>
                        <ButtonBase id={cur.name} onClick={() => {
                            if (props.onTileClick === undefined) return;
                            props.onTileClick(index);
                        }} style={{
                            width: "180px",
                            height: "110px"
                        }}>
                            <div className={classes.card}
                                 style={((index + 2) % 3 === 2) ? { marginLeft: 0 } : ((index + 2) % 3 === 1) ? {
                                     marginRight: 0,
                                     float: "right"
                                 } : { margin: "auto" }}>
                                {warning}
                                <div id={`${cur.name}_list_icon`} style={{ width: "100%", height: "80%", display: "grid" }}>
                                    {icon(cur.type)}
                                </div>
                                <Typography id={`${cur.name}_text`} className={classes.text}>
                                    {cur.name}
                                </Typography>
                            </div>
                        </ButtonBase>
                    </GridListTile>
                );
            })}
        </GridList>
    );
}
