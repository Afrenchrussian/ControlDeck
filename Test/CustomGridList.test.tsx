import React from "react";
import renderer from "react-test-renderer";
import CustomGridList  from "../src/Components/Multi/CustomGridList";
import { shallow } from "enzyme";

const testData = [
    {name: "test1", type: 0},
    {name: "test2", type: 1},
    {name: "test3", type: 2},
    {name: "test4", type: 3}
]

test("Custom Grid List renders correctly", () => {
    const list = renderer.create(<CustomGridList  data={testData}/>);
    expect(list.toJSON()).toMatchSnapshot();
});

test("Renders correct amount of tiles", () => {
    const list  = shallow(<CustomGridList data={testData}/>)
    expect(list.children().length).toBe(4)
})

test("Test OnClick for tiles", () => {
    const mockCallback = jest.fn((value) => value);
    const list = shallow(<CustomGridList data={testData} onTileClick={mockCallback}/>)
    list.find("#test1").simulate("click");
    expect(mockCallback.mock.calls.length).toBe(1)
    expect(mockCallback).toHaveBeenCalledWith(0)
})

test("Warning text shows or not", () => {
    const mockCallback = jest.fn((element) => (element.name === "test3")?"show_test_warning":"no_warning_test");
    const list = shallow(<CustomGridList data={testData} checkForWarnings={mockCallback}/>)
    expect(list.find("#test3_warning").childAt(0).name()).toBe("ErrorOutlineTwoToneIcon")
    expect(list.find("#test3_warning").prop('title')).toBe("show_test_warning")
    expect(list.find("#test1_warning").prop('title')).toBe("no_warning_test")
})

test("Check the correct icon is shown", () => {
    const list = shallow(<CustomGridList data={testData}/>)
    expect(list.find("#test2_list_icon").childAt(0).name()).toBe("PhoneIphoneIcon")
})

test("Check the name for each tile is shown",() => {
    const list = shallow(<CustomGridList data={testData}/>)
    expect(list.find("#test4_text").text()).toBe("test4")
})
