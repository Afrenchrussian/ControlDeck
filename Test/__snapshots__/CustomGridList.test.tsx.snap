// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Grid List renders correctly 1`] = `
<ul
  className="MuiGridList-root"
  style={
    Object {
      "margin": "auto",
      "marginTop": "30px",
      "width": "602px",
    }
  }
>
  <li
    className="MuiGridListTile-root"
    style={
      Object {
        "height": "135px",
        "padding": 2,
        "width": "33.333333333333336%",
      }
    }
  >
    <div
      className="MuiGridListTile-tile"
    >
      <button
        className="MuiButtonBase-root"
        disabled={false}
        id="test1"
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "height": "110px",
            "width": "180px",
          }
        }
        tabIndex={0}
        type="button"
      >
        <div
          className="makeStyles-card-1"
          style={
            Object {
              "marginLeft": 0,
            }
          }
        >
          <div
            id="test1_list_icon"
            style={
              Object {
                "display": "grid",
                "height": "80%",
                "width": "100%",
              }
            }
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root makeStyles-icon_large-3"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 1h-8C6.12 1 5 2.12 5 3.5v17C5 21.88 6.12 23 7.5 23h8c1.38 0 2.5-1.12 2.5-2.5v-17C18 2.12 16.88 1 15.5 1zm-4 21c-.83 0-1.5-.67-1.5-1.5s.67-1.5 1.5-1.5 1.5.67 1.5 1.5-.67 1.5-1.5 1.5zm4.5-4H7V4h9v14z"
              />
            </svg>
          </div>
          <p
            className="MuiTypography-root makeStyles-text-2 MuiTypography-body1"
            id="test1_text"
          >
            test1
          </p>
        </div>
      </button>
    </div>
  </li>
  <li
    className="MuiGridListTile-root"
    style={
      Object {
        "height": "135px",
        "padding": 2,
        "width": "33.333333333333336%",
      }
    }
  >
    <div
      className="MuiGridListTile-tile"
    >
      <button
        className="MuiButtonBase-root"
        disabled={false}
        id="test2"
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "height": "110px",
            "width": "180px",
          }
        }
        tabIndex={0}
        type="button"
      >
        <div
          className="makeStyles-card-1"
          style={
            Object {
              "margin": "auto",
            }
          }
        >
          <div
            id="test2_list_icon"
            style={
              Object {
                "display": "grid",
                "height": "80%",
                "width": "100%",
              }
            }
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root makeStyles-icon_small-4"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 1h-8C6.12 1 5 2.12 5 3.5v17C5 21.88 6.12 23 7.5 23h8c1.38 0 2.5-1.12 2.5-2.5v-17C18 2.12 16.88 1 15.5 1zm-4 21c-.83 0-1.5-.67-1.5-1.5s.67-1.5 1.5-1.5 1.5.67 1.5 1.5-.67 1.5-1.5 1.5zm4.5-4H7V4h9v14z"
              />
            </svg>
          </div>
          <p
            className="MuiTypography-root makeStyles-text-2 MuiTypography-body1"
            id="test2_text"
          >
            test2
          </p>
        </div>
      </button>
    </div>
  </li>
  <li
    className="MuiGridListTile-root"
    style={
      Object {
        "height": "135px",
        "padding": 2,
        "width": "33.333333333333336%",
      }
    }
  >
    <div
      className="MuiGridListTile-tile"
    >
      <button
        className="MuiButtonBase-root"
        disabled={false}
        id="test3"
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "height": "110px",
            "width": "180px",
          }
        }
        tabIndex={0}
        type="button"
      >
        <div
          className="makeStyles-card-1"
          style={
            Object {
              "float": "right",
              "marginRight": 0,
            }
          }
        >
          <div
            id="test3_list_icon"
            style={
              Object {
                "display": "grid",
                "height": "80%",
                "width": "100%",
              }
            }
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root makeStyles-icon_large-3"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M18.5 0h-14C3.12 0 2 1.12 2 2.5v19C2 22.88 3.12 24 4.5 24h14c1.38 0 2.5-1.12 2.5-2.5v-19C21 1.12 19.88 0 18.5 0zm-7 23c-.83 0-1.5-.67-1.5-1.5s.67-1.5 1.5-1.5 1.5.67 1.5 1.5-.67 1.5-1.5 1.5zm7.5-4H4V3h15v16z"
              />
            </svg>
          </div>
          <p
            className="MuiTypography-root makeStyles-text-2 MuiTypography-body1"
            id="test3_text"
          >
            test3
          </p>
        </div>
      </button>
    </div>
  </li>
  <li
    className="MuiGridListTile-root"
    style={
      Object {
        "height": "135px",
        "padding": 2,
        "width": "33.333333333333336%",
      }
    }
  >
    <div
      className="MuiGridListTile-tile"
    >
      <button
        className="MuiButtonBase-root"
        disabled={false}
        id="test4"
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "height": "110px",
            "width": "180px",
          }
        }
        tabIndex={0}
        type="button"
      >
        <div
          className="makeStyles-card-1"
          style={
            Object {
              "marginLeft": 0,
            }
          }
        >
          <div
            id="test4_list_icon"
            style={
              Object {
                "display": "grid",
                "height": "80%",
                "width": "100%",
              }
            }
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root makeStyles-icon_small-4"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M18.5 0h-14C3.12 0 2 1.12 2 2.5v19C2 22.88 3.12 24 4.5 24h14c1.38 0 2.5-1.12 2.5-2.5v-19C21 1.12 19.88 0 18.5 0zm-7 23c-.83 0-1.5-.67-1.5-1.5s.67-1.5 1.5-1.5 1.5.67 1.5 1.5-.67 1.5-1.5 1.5zm7.5-4H4V3h15v16z"
              />
            </svg>
          </div>
          <p
            className="MuiTypography-root makeStyles-text-2 MuiTypography-body1"
            id="test4_text"
          >
            test4
          </p>
        </div>
      </button>
    </div>
  </li>
</ul>
`;
